### ISSUE TITLE
Interactively choose transform operator for TransformStringByInput

### ISSUE THREAD

**Issue by t9md** [t9md/atom-vim-mode-plus#138]
In recent [commit](https://github.com/t9md/atom-vim-mode-plus/commit/5e538d771ccbe5bd293cf7b427c502ce728b6a3e)  I introduced TransformStringByInput.

Which allow user to select  transform to WHICH afterwards.
I want to experiment how it like if I add UI to select transforming operator interactively.
I mean allowing user to chose fuzzy key type to choose opearator like autocomplete-plus.

Currently user still have to remember **one character** that mapped to particular TransformString operator.

```coffeescript
class TransformStringByInput extends TransformString
  @extend()
  requireInput: true
  requireTarget: true

  transformingTable:
    'c': "CamelCase"
    '-': "DashCase"
    '_': "SnakeCase"
    'u': "LowerCase"
    'U': "UpperCase"
    '~': "ToggleCase"
    '%': "EncodeUriComponent"
    '@': "DecodeUriComponent" # FIXME for better nemonic

```

P.S.
For this `transformingTable` property, user can easily modify its table(add/remove entry), by following code.

```coffeescript
# init.coffee
# -------------------------
# General service consumer factory
getConsumer = (packageName, provider) ->
  (fn) ->
    atom.packages.onDidActivatePackage (pack) ->
      return unless pack.name is packageName
      service = pack.mainModule[provider]()
      fn(service)

# get vim-mode-plus service API provider
consumeVimModePlus = getConsumer 'vim-mode-plus', 'provideVimModePlus'

consumeVimModePlus ({Base}) ->
  TransformStringByInput = Base.getClass('TransformStringByInput')
  # E.g If I use only CamelCase and DashCase
  TransformStringByInput::transformingTable =
    'c': "CamelCase"
    '-': "DashCase"
```

**Comment by t9md** [comment]
![atom-vmp-transform-string-by-select-list](https://cloud.githubusercontent.com/assets/155205/12418281/caecab8c-bef1-11e5-9699-447e5b7ed0e9.gif)

Did experiment by dirty hacking vim-mode-plus-ex-mode.
I'm not sure this should fall into core though...

**Comment by t9md** [comment]
Further aggressive enhancement done.
Now dirtily replacing running operation by `vimState.reset()`, then `vimState.operationStack::run()` to support transformer like `Surround`, `SplitString` which require interactive user input.

**Comment by t9md** [comment]
@bronson, @chenglou, @crshd, @jordwalke

Do you guys have any opinion for this?
I want to know how other vmp user think of this feature.
Currently vmp support following common string transformation.
But keymap is limited(plus I'm not  providing default keymap for compatibility), remembering is sometime difficult for some users.
This feature allow user to select transformation operator(=transformer) from fuzzy-input select list.
Since its operator, once selected some transformer, user can repeat with `.` for subsequent transformation with same transformer against same target(word, inner-parenthesis etc.)

```coffeescript
  selectListItems: [
    "CamelCase"
    "DashCase"
    "SnakeCase"
    "LowerCase"
    "UpperCase"
    "ToggleCase"
    "EncodeUriComponent"
    "DecodeUriComponent"
    "JoinByInput"
    "JoinWithKeepingSpace"
    "Reverse"
    "SplitString"
    "Surround"
    "MapSurround"
  ]
```

**Comment by crshd** [comment]
These are nice features to have. However, I don't think they should be in the vmp core. These feel more like separate (non-vmp) packages, which exist already. Instead of duplicating features that can be found elsewhere, I'd rather see some improvements on the core (*cough* macros).

Except for surround. That is awesome, and I use it all the time. And I think the majority of vmp/vim users uses it as well. That should stay in core.

**Comment by t9md** [comment]
@crshd I got your opinion thanks!
But I want to clear misunderstanding of you, these transformation feature is **already implemented and provided** as vmp core.

What I'm asking here is specifically about implementing **selecting transformer from fuzzy-finder selectList** as vmp core feature.

**Comment by t9md** [comment]
One complement for Operator.

Although many no-vmp Atom package have similar feature for common string transformation, Implementing it as Vim's Operator is very important.

- It can repeat with `.`
- It can compose any kind of Motion, TextObject

Those are killer editing power of Vim, combination of Operator and Target(Motion and TextObject).

Yes, vim-surround atom package can surround, but it can't repeat as `.` since its not implemented as Operator.
In my opinion, any outer-vmp string transformation package need wrapper operator to get [true power of Vim](http://whileimautomaton.net/2008/11/vimm3/operator).

**Comment by crshd** [comment]
Oh, I know it's included already. I'm already using it. I'm saying it works OK as it is, and shouldn't have any more work done to it, unless you're splitting it into separate packages.

Personally, I don't care too much about `.`. These usually are actions I only need to run once. And thanks to muscle memory, I've typed `,siw"` before remembering I could have used `.`.

For me this falls under `atom-is-not-vim`. Yes, it would be nice to have operator everywhere, but Atom will stay Atom. You'd have to reinvent the wheel over and over again to integrate existing Atom plugins into vmp.

I believe vmp core should be focusing on vim core, and functionality that can be found in vim addons should be in separate vmp packages.

**Comment by t9md** [comment]
@crushd oh, OK. I misunderstood sorry.
Agree for keep core focusing important thing and let plugin handle nice-to-have feature. Once I wrote same sentence for vim-mode issue.
Since I forked and refactoring as vmp, I've implemented lots of features not fall into this rule.

As a maintainer I still experimenting how-it-like when implementing feature as vmp.

When I add feature as core the criteria is very loose.
If I believe it is simple and maintainable, I've experimentaly added.
I've wanted to postpone heavy feature like Macro because I don't want to be overwhelmed from bunch of user's request to ask 'macro not work for this operation'...
